### PR TITLE
improve union type checker error message

### PR DESCRIPTION
**Pull request by johann-sonntagbauer**
PropTypes type checking: Add additional check to UnionType checker to ensure only functions are used as argument of oneOfType.

This is especially handy to catch typos during PropType definition like for example:
```javascript
Greeting.propTypes = {
  name: React.PropTypes.oneOfType([React.PropTypes.sting, React.PropTypes.func])
};
```

The pull request will print a warning 
`Invalid argument supplied to oneOfType, expected an array containing check functions.`

Pull Request taken over from React main Repository
https://github.com/facebook/react/pull/8823

### CODE REVIEW

File: factory.js [hunk 1]
```diff
@@ -326,6 +326,13 @@ module.exports = function (isValidElement) {
       return emptyFunction.thatReturnsNull;
     }
 
+    for (var i = 0; i < arrayOfTypeCheckers.length; i++) {
+      if (typeof arrayOfTypeCheckers[i] !== 'function') {
+        warning(false, 'Invalid argument supplied to oneOfType, expected an array containing check functions.');
```
Comment (aweary) [inline factory.js]: What do you think about including the index and type in the warning?

> Invalid argument supplied to oneOfType, expected an array containing check functions but received null at index 2

Comment (johann-sonntagbauer) [inline factory.js]: good idea, will add the information

**Comment by johann-sonntagbauer** [comment]
@aweary introduced the change request of yours. Message is now much clearer.

**Comment by johann-sonntagbauer** [comment]
Tried to rebase onto current master. Hope I interpreted the changes in the right way. Please ping me when the master is stable enough to rebase again.

**Comment by johann-sonntagbauer** [comment]
@aweary thx for review. Changes are good.

**Pull request merged**